product-bundles-bulk-discounts-for-woocommerce.php

```diff
@@ -471,9 +471,12 @@ public static function bundled_cart_item_discount( $cart_item, $bundle ) {
 
 				$bundled_items_data = $container[ 'stamp' ];
 				$total_quantity     = 0;
+				$bundled_items      = $container[ 'data' ]->get_bundled_items();
 
-				foreach ( $bundled_items_data as $bundled_item_data ) {
-					if ( isset( $bundled_item_data[ 'quantity' ] ) ) {
+				foreach ( $bundled_items_data as $bundled_item_id => $bundled_item_data ) {
+					$bundled_item = $bundled_items[ $bundled_item_id ];
+
+					if ( isset( $bundled_item_data[ 'quantity' ] ) && ( ! $bundled_item->is_optional() || ( isset( $bundled_item_data[ 'optional_selected' ] ) && $bundled_item_data[ 'optional_selected' ] === 'yes' ) ) ) {
```

Comment:
It probably won't make much of a difference performance-wise, but how about changing this line to:

`if ( isset( $bundled_item_data[ 'quantity' ] ) && ( ! isset( $bundled_item_data[ 'optional_selected' ] ) || $bundled_item_data[ 'optional_selected' ] === 'yes' ) ) {`

Then L474 won't be needed. 

Again, bundled item objects are runtime-cached, so `get_bundled_items` doesn't really add any overhead -- but if everything we need is in the `$bundled_item_data` array, it wouldn't hurt to skip that `get_bundled_items` call.

Reply:
I am not entirely sure about how the main plugin handles this, but this might lead to issues in cases where the product is optional but the `$bundle_configuration` array does not contain a `optional_selected` key. The product would then be added to the discount quantity. Now whether this is correct is obviously up to you guys, but I would have thought that `optional_selected` would default to `no` (and thus the product not being added).

Reply:
> in cases where the product is optional but the `$bundle_configuration` array does not contain a `optional_selected` key.

Hmmm... this should never happen with posted data. If a bundled item is optional and the config array is manually supplied (for example via custom code), then it's up to the developer to supply correct data.

The integrity of the config array is always validated on the `woocommerce_check_cart_items` hook -- so if a config is "stale" you'll get an error in the cart, and will not be able to proceed to checkout.